crates/evm/src/precompile_cache.rs

```diff
@@ -0,0 +1,341 @@
+    precompile_cache_misses: metrics::Gauge,
+
+    /// Precompile cache size
+    ///
+    /// NOTE: this uses the moka caches' `entry_count`, NOT the `weighted_size` method to calculate
+    /// size.
+    precompile_cache_size: metrics::Gauge,
+}
+
+/// A custom precompile provider that wraps a precompile provider and potentially a cache for it.
+#[derive(Clone, Debug)]
+pub struct MaybeCachedPrecompileProvider<P> {
+    /// The precompile provider to wrap.
+    precompile_provider: P,
+    /// The cache to use.
+    #[cfg(feature = "std")]
+    cache: Option<alloc::sync::Arc<PrecompileCache>>,
+    /// The spec id to use.
+    spec: SpecId,
+    /// Cache metrics.
+    #[cfg(feature = "metrics")]
+    metrics: CachedPrecompileMetrics,
+}
+
+impl<P> MaybeCachedPrecompileProvider<P> {
+    /// Given a [`PrecompileProvider`]  and cache for a specific precompile provider,
+    /// create a cached wrapper that can be used inside Evm.
+    #[cfg(feature = "std")]
+    pub fn new_with_cache(
+        precompile_provider: P,
+        cache: alloc::sync::Arc<PrecompileCache>,
+    ) -> Self {
+        Self {
+            precompile_provider,
+            #[cfg(feature = "std")]
+            cache: Some(cache),
+            spec: Default::default(),
+            #[cfg(feature = "metrics")]
+            metrics: Default::default(),
+        }
+    }
+
+    /// Creates a new `MaybeCachedPrecompileProvider` with cache disabled.
+    pub fn new_without_cache(precompile_provider: P) -> Self {
+        Self {
+            precompile_provider,
+            #[cfg(feature = "std")]
+            cache: None,
+            spec: Default::default(),
+            #[cfg(feature = "metrics")]
+            metrics: Default::default(),
+        }
+    }
+}
+
+impl<CTX: ContextTr, P: PrecompileProvider<CTX, Output = InterpreterResult>> PrecompileProvider<CTX>
+    for MaybeCachedPrecompileProvider<P>
+{
+    type Output = P::Output;
+
+    fn set_spec(&mut self, spec: <CTX::Cfg as Cfg>::Spec) -> bool {
+        self.precompile_provider.set_spec(spec.clone());
+        self.spec = spec.into();
+        true
+    }
+
+    #[cfg(feature = "std")]
+    fn run(
+        &mut self,
+        context: &mut CTX,
+        address: &Address,
+        inputs: &InputsImpl,
+        is_static: bool,
+        gas_limit: u64,
+    ) -> Result<Option<Self::Output>, String> {
+        use revm::interpreter::{Gas, InstructionResult};
+
+        if let Some(cache) = &self.cache {
+            let key =
+                PrecompileKey { address: *address, spec: self.spec, input: inputs.input.clone() };
+
+            let cache_result = cache.cache.get(&key);
+
+            if let Some(ref entry) = cache_result {
+                // if gas_limit is below known lower bound, we know it will fail with OOG
+                if gas_limit <= entry.lower_gas_limit {
+                    #[cfg(feature = "metrics")]
+                    self.metrics.precompile_cache_hits.increment(1);
```

Comment:
I'd like to avoid all these extra cfgs in this function and instead extract this to a non feature gated function that then has the feature gated body, this would make this function easier to read